### DISCUSSION TITLE
Player object listed in room description (and in the pane) en Español

### DISCUSSION THREAD

**Discussion by KVonGit** [textadventures/quest#1283] [Creating games with Quest] [answered]
```xml
<asl version="550" template="Español">

  <include ref="Espanol.aslx"/>
  <include ref="Core.aslx"/>

  <game name="I Can See Myself in Spanish">
    <gameid>7dd4e4d8-80f6-4371-bbf5-d35f56b6c09c</gameid>
    <version>1.0</version>
    <firstpublished>2024</firstpublished>
  </game>

  <object name="lugar">
    <inherit name="editor_room" />
    <isroom />

    <object name="jugador">
      <inherit name="editor_object" />
      <inherit name="editor_player" />
    </object>
  </object>

</asl>
```

---
![image](https://github.com/user-attachments/assets/b3dbbbfd-c5ce-4d1c-b81a-6025ca9fef12)

---
> [!NOTE]
> This is a fresh game created in 5.8 in Spanish, with no changes to the code before playing. The ASL version was not changed to 580 in **Espanol.aslx** during the update to Quest 5.8 (it was overlooked along with a few others), but changing it during testing now does not effect this issue. (ALSO NOTE: The ASL version is changed to 580 automatically whenever the editor saves.)

---
I notice the `jugador` object inherits the same two types as the `player` object in English, but they are both "dummy" editor types which won't matter once the game is published. That is moot here, because I am playing the game directly from the editor; so, those types do effect things the way I tested this.

---
**[ScopeReachableNotHeldForRoom](https://github.com/textadventures/quest/blob/7a0376151cac36728157d2a49827e783777e8c7e/WorldModel/WorldModel/Core/CoreScopes.aslx#L119C1-L134)**
```C#
    result = NewObjectList()
    foreach (obj, GetAllChildObjects(room)) {
      if (ContainsReachable(room, obj) and obj <> game.pov and not Contains(game.pov, obj)) {
```

---
🤔 

![image](https://github.com/user-attachments/assets/d4c9b287-4151-425f-a967-19dae4414a34)

---
Oh... That's not me. `game.pov` is still set to `player`.

If I create another room to go to, I leave my Spanish counterpart behind.

![image](https://github.com/user-attachments/assets/41858b89-cfb4-4687-b8f7-089e8f56f9e7)

---
Okay... In [`StartGame`](https://github.com/textadventures/quest/blob/7a0376151cac36728157d2a49827e783777e8c7e/WorldModel/WorldModel/Core/Core.aslx#L252-L266), if `game.pov` is not set, it will create a `player` object and set `game.pov` to that.

So, when we make games in a language other than English, we have to set `game.pov` to the existing default player object that shows up in the editor in `game.start`?

Should we just always use `player` as the object's name and change the alias for languages other than English?

I can't find anything in the docs explaining that `game.pov` needs to be set in the start script. There's no warning in the editor, either.

Is this an issue, or am I overlooking something? (Or does it say it in the editor, and I just can't read the language? (I barely understand Spanish.))

---
Adding this to my start script:
```c#
game.pov = jugador
```

![image](https://github.com/user-attachments/assets/0f666df5-5f34-4b5f-a469-0d468baaa874)

---
Oh... Quest is cheeky today!

Note the lack of a `player` object in my code:

<details><summary>Click here to view code.</summary>

```xml
<!--Saved by Quest 5.8.6836.13983-->
<asl version="580">
  <include ref="Espanol.aslx" />
  <include ref="Core.aslx" />
  <game name="I Can See Myself in Spanish">
    <gameid>7dd4e4d8-80f6-4371-bbf5-d35f56b6c09c</gameid>
    <version>1.0</version>
    <firstpublished>2024</firstpublished>
    <start type="script">
      game.pov = jugador
    </start>
  </game>
  <object name="lugar">
    <inherit name="editor_room" />
    <isroom />
    <description type="script">
      msg ("")
      msg ("ScopeReachableNotHeldForRoom(lugar):")
      msg (ScopeReachableNotHeldForRoom(lugar))
      msg ("")
      msg ("game.pov:")
      msg (game.pov)
    </description>
    <object name="jugador">
      <inherit name="editor_object" />
      <inherit name="editor_player" />
    </object>
    <exit alias="norte" to="room2">
      <inherit name="northdirection" />
    </exit>
  </object>
  <object name="room2">
    <inherit name="editor_room" />
    <exit alias="sur" to="lugar">
      <inherit name="southdirection" />
    </exit>
  </object>
</asl>
```

</details>

---
That's because [that bit of code in `StartGame`](https://github.com/textadventures/quest/blob/7a0376151cac36728157d2a49827e783777e8c7e/WorldModel/WorldModel/Core/Core.aslx#L252-L276) runs before `game.start` and before any `initialise` scripts on any objects.

So, I must put this in `game.inituserinterface` then, it seems....

...but what if a non-English game can switch POV during play??? This would reset to this first POV object when resuming a saved game!

I can only think of two ways to fix this:
- the `player` object is always the starting POV, and the alias should be changed in games that are not English
- `StartGame` should be modified to handle this
  - either by switching the order around during this part, or by searching all objects for allowed POV objects

It seems like making everyone keep the object named `player` and having them change the alias would be the best/easiest solution. Isn't "player" one of those words everyone around the world recognizes?

---
I ended up doing this:

```xml
    <inituserinterface type="script">
      if (not HasAttribute(game, "pov")) {
        game.pov = jugador
      }
    </inituserinterface>
```

---
Am I completely missing something?

**Comment by angelwedge** [comment, marked as answer]
I think that the issue is in `Core.aslx`.

We have the lines (line 252 onwards):
```
    if (game.pov = null) {
      playerObject = GetObject("player")
      if (playerObject = null) {
```
Really, I suspect that this default name should be checked against some name taken from a template specified in the language file, trying "player" as a backup only if that one fails.

Also, it would probably make sense if `WorldModel/WorldModel/Core/Templates/Espanol.template` included a default value for the `pov` attribute.

**Reply by KVonGit** [reply]
> Also, it would probably make sense if WorldModel/WorldModel/Core/Templates/Espanol.template included a default value for the pov attribute.

This sounds like the best solution to me.

**Reply by ghost** [reply]
Every other template besides **Espanol.template** has `room` and `player`.

I'm just changing this file to match the rest.
- `room` should be renamed anyway
- `player` never prints during play (unless POV is switched, and, well that's besides the point -- the author who messes with POV is in charge of all that)

---
If the other templates weren't already using `player`, I would add the suggested fix to my PR. (Sorry, I only tested this in Spanish before creating this thread, because I can understand it better than the other languages.)

---
NOTE:

I see how this can still be problematic if the author changes the name of the `player` object, and I now agree that your first thought is probably the best way to cover all the bases (as long as the author knows the proper procedure):

> Really, I suspect that this default name should be checked against some name taken from a template specified in the language file, trying "player" as a backup only if that one fails.

---
I'm just changing this one template for now, though, unless Alex just wants to modify `StartGame` in this update. We've already got a ton of other changes submitted. `:)`

**Comment by KVonGit** [comment]
Fixed in #1238
